seccomp/umount.go

```diff
@@ -117,15 +124,81 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
 	return u.tracer.createContinueResponse(u.reqId), nil
 }
 
+// processUmmountAllowed purpose is to prevent immutable resources from being
+// unmounted.
+//
+// Method will return 'true' if the umount operation is deemed legit, and will
+// return 'false' otherwise.
+func (u *umountSyscallInfo) processUmountAllowed(
+	mip domain.MountInfoParserIface) (bool, *sysResponse) {
+
+	// There must be mountinfo state present for this target. Otherwise, let
+	// kernel handle the error.
+	info := mip.GetInfo(u.Target)
+	if info == nil {
+		return true, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	// Return 'eperm' if the umount operation is over an immutable mountpoint.
+	if u.cntr.IsImmutableMountID(info.MountID) {
+		return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+	}
+
+	// At this point we have already concluded that the mountpoint being
+	// unmounted doesn't match any of the immutable resources, so we could
+	// be tempted to allow all umount instructions reaching this stage to
+	// succeed. However, there is one particular case that we must identify
+	// to prevent the exposure of immutable masked resources. This is the case
+	// of an umount instruction launched from an unshare() context that targets
+	// an immutable resource. The rest of the code in this function deals with
+	// the identification and handling of this particular case.
+
+	p := u.processData
+
+	// Allow umount if this one is launched from a process whose root-inode
+	// differs from the sys container's initPid one.
+	if p.RootInode() != u.cntr.InitProc().RootInode() {
```

Comment:
We have a bug here: just because the root is different, it does not mean it's not referencing the same mount. For example, say a process inside the sys container is in a dedicated mount namespace has a root of /var/lib/; if that process decides to unmount /docker, it's really unmounting /var/lib/docker inside the sys container, so it should not be allowed (since /var/lib/docker is an immutable mount). But this code will allow it.

Reply:
The bug is present. It occurs when a process unshares its mount ns, does a chroot to `/var/lib/` and unmounts `/docker`.

In this case the `IsImmutableMountID()` check above won't catch it (due to the new mount namespace) and since the process did a chroot, this code will allow the umount to succeed, when it shouldn't have.

Reply:
It seems to me you want to first check if the process doing the umount is in the same mount namespace as the sys container's root process. 

If it is, then the IsImmutableMountID() check suffices. 

If it is not, then we can't check by mount-ID, so we have to check by other means. Currently the code checks by umount target path, but in order to do that you have to know the path with respect to the sys container's root. I don't see the code doing that. In fact we can't do that as I mentioned in another comment, as we don't have enough info from the kernel to convert the umount target path to a path relative to the container's root. The only thing that comes to mind is that we check by umount target path inode. 

In addition, we have to check if there is a mount on top of the immutable mount. If there is, the umount should be allowed to proceed. Function `IsRecursiveBindMount` attempts to do that, but it's not catching non-bind mounts (e.g., a tmpfs mount on top of an immutable mount for example).